src/cocotb/clock.py

```diff
@@ -109,52 +115,61 @@ async def custom_clock():
     """
 
     def __init__(
-        self, signal, period: Union[float, Fraction, Decimal], units: str = "step"
+        self,
+        signal,
+        period: Union[float, Fraction, Decimal],
+        units: str = "step",
+        impl: str = "auto",
     ):
         self.signal = signal
         self.period = get_sim_steps(period, units)
-        self.half_period = get_sim_steps(period / 2, units)
         self.frequency = 1 / get_time_from_sim_steps(self.period, units="us")
-        self.hdl = None
-        self.signal = signal
-        self.coro = None
-        self.mcoro = None
+        self.t_high = self.period // 2
+
+        # If not using GpiClock, clkobj is None
+        if impl == "auto":
+            impl = "gpi" if cocotb._conf.trust_inertial else "py"
+        if impl == "gpi":
+            self.clkobj = clock_create(self.signal._handle)
```

Comment:
Namely here.

The issue is that if the `Clock` object goes out of scope, the `GpiClock` will be stopped and cleaned up as well. However, `Clock` currently allows itself to leak the actual clock driving task and this constitutes a change in behavior.

I think the easiest way to make it work like it used to is by wrapping it in a Task.

```python
self.clkobj = cocotb.start_soon(self._gpi_clock_runner())
```

```python
async def _gpi_clock_runner(self) -> None:
    gpi_clock = clock_create(self.signal._handle)
    await Event().wait() # Forever
    # when task is killed, gpi_clock will be cleaned up, stopping the clock
```

But more ideally once CancelledError works.
```python
async def _gpi_clock_runner(self) -> None:
    gpi_clock = clock_create(self.signal._handle)
    try:
        await Event().wait() # Forever
    finally:
        gpi_clock.stop()
```

Reply:
Nvm, you are already doing that.